File: src/main/java/core/procedures/io/CallWithInputFile.java
package core.procedures.io;

import core.exceptions.SCMFileNotFoundException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.procedures.IFn;
import core.scm.Cons;
import core.scm.InputPort;
import core.scm.Thunk;

import java.io.FileInputStream;
import java.io.FileNotFoundException;

public final class CallWithInputFile extends AFn {

  public CallWithInputFile() {
    super(new FnArgsBuilder().min(2).max(2).mandatory(new Class[]{CharSequence.class, IFn.class}).build());
  }

  @Override
  public String getName() {
    return "call-with-input-file";
  }

  @Override
  public Object apply(Object... args) {
    String filename = args[0].toString();
    InputPort inputPort;
    try {
      inputPort = new InputPort(new FileInputStream(filename));
    } catch (FileNotFoundException e) {
      throw new SCMFileNotFoundException(filename);
    }
    IFn proc = ((IFn)args[1]);
    Cons sexp = Cons.list(proc, inputPort);
    return new Thunk(sexp);
  }
}

File: src/main/java/core/procedures/io/CallWithInputFile.kt
package core.procedures.io

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.procedures.IFn
import core.scm.InputPort
import core.scm.Thunk
import core.scm.specialforms.Try

import java.io.FileInputStream

class CallWithInputFile : AFn<Any?, Any>(name = "call-with-input-file", arity = Exactly(2),
                                         mandatoryArgsTypes = arrayOf(CharSequence::class.java, IFn::class.java)) {

    /* (try (proc in) (finally (close-input-port in)))*/
    override operator fun invoke(arg1: Any?, arg2: Any?) = InputPort(FileInputStream(arg1!!.toString())).let {
        Thunk(listOf(Try, listOf(arg2, it),
              listOf(Try.FINALLY, listOf(CloseInputPort(), it))))
    }
}
